src/kubelet/client/client.go

```diff
@@ -0,0 +1,158 @@
+package client
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"path"
+	"time"
+
+	"github.com/newrelic/infra-integrations-sdk/log"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+
+	"github.com/newrelic/nri-kubernetes/v2/internal/config"
+	"github.com/newrelic/nri-kubernetes/v2/src/client"
+	"github.com/newrelic/nri-kubernetes/v2/src/prometheus"
+)
+
+const (
+	healthzPath             = "/healthz"
+	defaultHTTPKubeletPort  = 10255
+	defaultHTTPSKubeletPort = 10250
+	defaultTimeout          = time.Millisecond * 5000
+)
+
+// Client implements a client for Kubelet, capable of retrieving prometheus metrics from a given endpoint.
+type Client struct {
+	// TODO: Use a non-sdk logger
+	logger        log.Logger
+	doer          client.HTTPDoer
+	endpoint      url.URL
+	apiServerHost string
+}
+
+type OptionFunc func(kc *Client) error
+
+// WithLogger returns an OptionFunc to change the logger from the default noop logger.
+func WithLogger(logger log.Logger) OptionFunc {
+	return func(kc *Client) error {
+		kc.logger = logger
+		return nil
+	}
+}
+
+// New builds a Client using the given options.
+func New(kc kubernetes.Interface, config config.Mock, inClusterConfig *rest.Config, opts ...OptionFunc) (*Client, error) {
+	c := &Client{
+		logger:        log.New(false, io.Discard),
+		apiServerHost: inClusterConfig.Host,
+	}
+
+	for i, opt := range opts {
+		if err := opt(c); err != nil {
+			return nil, fmt.Errorf("applying option #%d: %w", i, err)
+		}
+	}
+
+	conn, err := c.setupConnection(kc, tripperWithBearerToken(inClusterConfig.BearerToken), config)
+
+	if err != nil {
+		return nil, fmt.Errorf("connecting to kubelet: %w", err)
+	}
+
+	c.doer = conn.client
+	c.endpoint = conn.url
+
+	return c, nil
+}
+
+func (c *Client) setupConnection(kc kubernetes.Interface, tripperBearerToken http.RoundTripper, config config.Mock) (*connParams, error) {
+	kubeletPort, err := getKubeletPort(kc, config.NodeName)
+	if err != nil {
+		return nil, fmt.Errorf("getting kubelet port: %w", err)
+	}
+
+	conn, err := c.setupLocalConnection(tripperBearerToken, config.NodeIP, kubeletPort)
+	if err == nil {
+		c.logger.Debugf("connected to Kubelet directly with nodeIP")
+		return conn, nil
+	}
+
+	c.logger.Debugf("Kubelet connection with nodeIP failed: %v", err)
+	c.logger.Debugf("Connecting to kubelet directly with API proxy")
```

Comment:
These messages might be relevant for the user, perhaps we could word them slighly less scary (do not say "failed") and bump the log level to info.